### PR TITLE
Handle ProcessExit event

### PR DESCRIPTION
**Pull request by lostmsu**
This avoids segfault when CLR (in particular Mono) is unloaded before Python stops.

### What does this implement/fix? Explain your changes.

1. Added a handler to `AppDomain.ProcessExit`, that fires in the situations like this instead of `AppDomain.DomainUnload`. This allows Python.NET to remove all slots, that are implemented in C#.
2. When initialized as extension (e.g.  Python hosts CLR), release GIL acquired during shutdown so that Python finalization can resume.

### Does this close any currently open issues?

This is (at least one of) the reason of CI test crashes in https://github.com/pythonnet/pythonnet/pull/1134

### CODE REVIEW

File: src/runtime/runtime.cs [hunk 1]
```diff
@@ -393,6 +393,10 @@ internal static void Shutdown(ShutdownMode mode)
                 {
                     Py_Finalize();
                 }
+                else
+                {
+                    PyGILState_Release(state);
```
Comment (filmor) [inline src/runtime/runtime.cs:398]: This change seems unrelated, why did you add this?

Comment (lostmsu) [reply src/runtime/runtime.cs:398]: Without this `Py_Finalize` called when Python finally shuts down hangs because it can't acquire GIL.

**Comment by lostmsu** [comment]
I debugged the test failures, and they appear to be caused by gaps in domain reload support: when .NET objects that Python sees get severed from CLR, subsequent calls to their `tp_dealloc` fail because `GetManagedObject` returns `null`.

**Comment by lostmsu** [comment]
Found a way to solve the same problem and more without handling `ProcessExit`

**Comment by lostmsu** [comment]
Or maybe not. Seems to still be essential for https://github.com/pythonnet/pythonnet/pull/1134
